molpipeline/mol2mol/filter.py

```diff
@@ -120,38 +164,207 @@ def pretransform_single(self, value: RDKitMol) -> OptionalMol:
+        to_process_value = (
+            Chem.AddHs(value) if 1 in self.allowed_element_numbers else value
+        )
+
+        elements_list = [atom.GetAtomicNum() for atom in to_process_value.GetAtoms()]
+        elements_counter = Counter(elements_list)
+        if any(
+            element not in self.allowed_element_numbers for element in elements_counter
+        ):
             return InvalidInstance(
-                self.uuid,
-                f"Molecule contains following forbidden elements: {forbidden_elements}",
-                self.name,
+                self.uuid, "Molecule contains forbidden chemical element.", self.name
             )
+        for element, (min_count, max_count) in self.allowed_element_numbers.items():
+            count = elements_counter[element]
+            if (min_count is not None and count < min_count) or (
+                max_count is not None and count > max_count
+            ):
+                return InvalidInstance(
+                    self.uuid,
+                    f"Molecule contains forbidden number of element {element}.",
+                    self.name,
+                )
         return value
 
 
-class MixtureFilter(_MolToMolPipelineElement):
-    """MolToMol which removes molecules composed of multiple fragments."""
+# should we combine smarts and smiles filter within a single class? option usesmiles?
+# should we check the input patterns for valid smarts/smiles?
+# should we apply the same logic to ElementFilter?
+class SmartsFilter(_BasePatternsFilter):
+    """Filter to keep or remove molecules based on SMARTS patterns."""
+
+    def _calculate_single_element_value(
+        self, filter_element: Any, value: RDKitMol
+    ) -> float:
+        """Calculate a single smarts match count for a molecule.
+
+        Parameters
+        ----------
+        filter_element: Any
+            smarts to calculate match count for.
+        value: RDKitMol
+            Molecule to calculate smarts match count for.
+
+        Returns
+        -------
+        float
+            smarts match count value.
+        """
+        return len(value.GetSubstructMatches(Chem.MolFromSmarts(filter_element)))
+
+
+class SmilesFilter(_BasePatternsFilter):
+    """Filter to keep or remove molecules based on SMILES patterns."""
 
-    def __int__(
+    def _calculate_single_element_value(
+        self, filter_element: Any, value: RDKitMol
+    ) -> float:
+        """Calculate a single smiles match count for a molecule.
+
+        Parameters
+        ----------
+        filter_element: Any
+            smiles to calculate match count for.
+        value: RDKitMol
+            Molecule to calculate smiles match count for.
+
+        Returns
+        -------
+        float
+            smiles match count value.
+        """
+        return len(value.GetSubstructMatches(Chem.MolFromSmiles(filter_element)))
+
+
+class DescriptorsFilter(_BaseKeepMatchesFilter):
```

Comment:
I would rather call the class `RDKitDescriptorsFilter` instead of `DescriptorsFilter` because it specifically works with RDKit's `Descriptors` module